CE: Add Backend API documentation (#1414)
vaadin/docs: articles/ce/advanced/backend-api.asciidoc

<3> We add an item listener to the Hazelcast list to handle new items and removed ones.
<4> Then all past events are initially delivered.
<5> Finally we return a [classname]`Registration` that can be used to remove the subscriber.

Let's now dive in into the custom method that effectively deliver the events, the [methodname]`deliverEvents` method.
This is a `synchronized` method to prevent it to be invoked by multiple threads simultaneously and avoid duplicate event to be notified to the subscriber.
The method keeps track of the Hazelcast list index to identify the next event and increments this index until all the events are delivered.
If an event identifier has been set as the starting point, no events are delivered until that identifier is reached.

[source,java]
----
include::{root}/src/main/java/com/vaadin/demo/ce/hazelcast/HazelcastBackend.java[tags=deliver-events,indent=0]
----

Finally, the last method to implement for the [classname]`EventLog` interface is the [methodname]`truncate` method.
This method serves the purpose to limit the number of events contained in the log, to avoid it to grow infinitely.

> It takes the identifier of the older know event that should be preserved, or — if a `null` identifier is provided — it empties the whole log.

Reviewer: ```suggestion
It takes the identifier of the oldest known event that should be preserved, or — if a `null` identifier is provided — it empties the whole log.
```

> To implement this behavior for Hazelcast, we create a simple [classname]`Predicate` and pass it to the list [methodname]`removeIf` method.

Reviewer: ```suggestion
To implement this behavior for Hazelcast, we create a simple [classname]`Predicate` and pass it to the list's [methodname]`removeIf` method.
```

[source,java]
----
include::{root}/src/main/java/com/vaadin/demo/ce/hazelcast/HazelcastBackend.java[tags=truncate,indent=0]
----

== Opening an Event Log

> Now that we have an Hazelcast implementation of the [classname]`EventLog` interface, we need to extends the [classname]`Backend` class to get its instances.

Reviewer: ```suggestion
Now that we have a Hazelcast implementation of the [classname]`EventLog` interface, we need to extend the [classname]`Backend` class to get its instances.
```

In addition, rephrase to clarify the meaning of "get its instances".

Author: Rephrased to:
> Now that we have a Hazelcast implementation of the `EventLog` interface, to be able to create and get instances of it we need to extend the `Backend` class.

> Since the implementation only depends on a single Hazelcast [classname]`IList`, it is very easy to implement the [methodname]`openEventLog` method and we do so by returning a new instance of the [classname]`HazelcastEventLog` with the list named after the `logId` parameter.

Reviewer: ```suggestion
Since our implementation only depends on a single Hazelcast [classname]`IList`, it is very easy to implement the [methodname]`openEventLog` method and we do so by returning a new instance of the [classname]`HazelcastEventLog` with the list named after the `logId` parameter.
```

[source,java]
----
include::{root}/src/main/java/com/vaadin/demo/ce/hazelcast/HazelcastBackend.java[tags=openEventLog,indent=0]
----

== Managing Event Log Snapshots

A _snapshot_ is an opaque representation of data at a certain moment in time.

> It can be used for example by nodes joining a cluster to quickly catch-up with a recent state of data without the need to reply all the events from the beginning of time.

Reviewer: ```suggestion
It can be used, for example, by nodes joining a cluster to quickly catch-up with a recent state of data without the need to replay all the events from the beginning of time.
```

> Snapshots are identified by a name and each version of a named snapshot is assigned with an additional unique identifier.

Reviewer: ```suggestion
Snapshots are identified by a name, and each version of a named snapshot is assigned with an additional unique identifier.
```

To implement this method for Hazelcast, we need some logic to verify that the latest snapshot is the expected one.

[source,java]
----
include::{root}/src/main/java/com/vaadin/demo/ce/hazelcast/HazelcastBackend.java[tags=replaceSnapshot,indent=0]
----

== Node identity and Membership events

The primary purpose of the Backend API is to support collaboration in applications deployed in clustered environments.
Every [classname]`Backend` instance represents a member of the cluster and is uniquely identified by a [classname]`UUID` which should be returned by the [methodname]`getNodeId` method.
For our Hazelcast implementation, we simply return the local member identifier.

[source,java]
----
include::{root}/src/main/java/com/vaadin/demo/ce/hazelcast/HazelcastBackend.java[tags=getNodeId,indent=0]
----

When multiple [classname]`Backend` instances are involved, it is necessary to know when a they join or leave the cluster.
For this purpose, the [classname]`Backend` should provide an implementation of the [methodname]`addMembershipListener` method that takes a [classname]`MembershipListener` and notifies it when cluster members join or leave.

> Hazelcast comes itself with the same concept, so the implementation is very straightforward: we only need to map Hazelcast's events to Collaboration Engine's [classname]`MembershipEvent` events, which take the [classname]`MembershipEventType` (`JOIN` or `LEAVE`) and the identifier of the member.

Reviewer: ```suggestion
Hazelcast itself uses with the same concept, so the implementation is very straightforward: we only need to map Hazelcast's events to Collaboration Engine's [classname]`MembershipEvent` events, which take the [classname]`MembershipEventType` (`JOIN` or `LEAVE`) and the identifier of the member.
```
